packages/server/src/unstable-core-do-not-import/router.ts

```diff
@@ -238,11 +257,20 @@ export function callProcedure(
 export function createCallerFactory<TRoot extends AnyRootTypes>() {
   return function createCallerInner<TRecord extends RouterRecord>(
     router: Router<TRoot, TRecord>,
+    options?: {
+      onError?: RouterCallerErrorHandler<TRoot['ctx']>;
+    },
   ): RouterCaller<TRoot, TRecord> {
     const _def = router._def;
+    const factoryOptions = options;
     type Context = TRoot['ctx'];
 
-    return function createCaller(maybeContext) {
+    return function createCaller(
+      maybeContext,
+      options?: {
+        onError?: RouterCallerErrorHandler<Context>;
+      },
```

Comment:
why do we need this option in both places?

Reply:
I figured that a handler on the factory itself would be useful for keeping things DRY when callers are being created in multiple places with the same error handler and from the same factory.
Although I just realized that throwing anything in the handler provided to createCallerFactory would prevent any other handler from overriding that behaviour in the first place, which kind of defeats the purpose.
I ended up writing a solution which passes any handler provided to createCallerFactory to the one provided to createCaller, middleware-style, but perhaps this complicates things more than necessary.
So adding the option only to createCaller might be the way to go, even if it could mean more repitition in some cases; it would definitely be easier to reason about.

Reply:
Yea if people are creating callers in multiple places, they could just have a function that they import and pass, since it's all one signature anyways...

```ts
const myErrorHandler: RouterErrorHandler<Context> = (opts) => { ...}

const createCaller = t.createCallerFactory(appRouter);

const caller = createCaller(context, { onError: myErrorHandler });
```

which doesn't really break DRY too much...

I wonder now though what the usecase is for declaring different error handlers at all? Wouldn't a root one be sufficient?

```ts
const createCaller = t.createCallerFactory(appRouter, { onError: myErrorHandler });
const caller1 = createCaller(context);
```

Reply:
Here's a branch with the continuation-passing I mentioned: https://github.com/jonnelaaglandwinder/trpc/tree/routercaller-onerror-next
It allows for defining handlers on both createCallerFactory and createCaller in an extendable manner, but I think it's a bit too convoluted.
I think defining the error handler in one place and then passing it to createCaller as you suggested strikes the right balance. It's somewhat more verbose and any resulting code would take more work to refactor, but that shouldn't be much of a problem with modern tooling. If any given choice is between DRY and KISS/YAGNI, the latter is probably preferable.
This branch only has the option on createCaller: https://github.com/jonnelaaglandwinder/trpc/tree/routercaller-onerror-createcaller-only